Sources/CNIOLinux/liburing_shims.c

```diff
@@ -0,0 +1,452 @@
+typedef int (*__io_uring_get_cqe_fp)(struct io_uring *ring,
+                  struct io_uring_cqe **cqe_ptr, unsigned submit,
+                  unsigned wait_nr, sigset_t *sigmask);
+
+// local static struct holding resolved function pointers from dlsym
+static struct _liburing_functions_t
+{
+    io_uring_get_probe_ring_fp io_uring_get_probe_ring;
+    io_uring_get_probe_fp io_uring_get_probe;
+    io_uring_free_probe_fp io_uring_free_probe;
+    io_uring_queue_init_params_fp io_uring_queue_init_params;
+    io_uring_queue_init_fp io_uring_queue_init;
+    io_uring_queue_mmap_fp io_uring_queue_mmap;
+    io_uring_ring_dontfork_fp io_uring_ring_dontfork;
+    io_uring_queue_exit_fp io_uring_queue_exit;
+    io_uring_peek_batch_cqe_fp io_uring_peek_batch_cqe;
+    io_uring_wait_cqes_fp io_uring_wait_cqes;
+    io_uring_wait_cqe_timeout_fp io_uring_wait_cqe_timeout;
+    io_uring_submit_fp io_uring_submit;
+    io_uring_submit_and_wait_fp io_uring_submit_and_wait;
+    io_uring_get_sqe_fp io_uring_get_sqe;
+    io_uring_register_buffers_fp io_uring_register_buffers;
+    io_uring_unregister_buffers_fp io_uring_unregister_buffers;
+    io_uring_register_files_fp io_uring_register_files;
+    io_uring_unregister_files_fp io_uring_unregister_files;
+    io_uring_register_files_update_fp io_uring_register_files_update;
+    io_uring_register_eventfd_fp io_uring_register_eventfd;
+    io_uring_register_eventfd_async_fp io_uring_register_eventfd_async;
+    io_uring_unregister_eventfd_fp io_uring_unregister_eventfd;
+    io_uring_register_probe_fp io_uring_register_probe;
+    io_uring_register_personality_fp io_uring_register_personality;
+    io_uring_unregister_personality_fp io_uring_unregister_personality;
+    __io_uring_sqring_wait_fp __io_uring_sqring_wait;
+    __io_uring_get_cqe_fp __io_uring_get_cqe;
+} liburing_functions;
+
+// Convenience macro for resolving
+#define _DL_RESOLVE(symbol) \
+    liburing_functions.symbol = (symbol ## _fp) dlsym(dl_handle, #symbol);  \
+    if ((err = dlerror()) != NULL) {  \
+        printf("WARNING: Failed to resolve " #symbol " from liburing, falling back on epoll()\n");  \
+        (void) dlclose(dl_handle); \
+        return -1;  \
+    }
+
+// dynamically load liburing and resolve symbols. Should be called once before using io_uring.
+// returns 0 on successful loading and resolving of functions, otherwise error
+
+// getting kernel version, just adopted from SO answer.
+int _check_compatible_kernel_version() {
+    struct utsname buffer;
+    char *p;
+    long ver[16];
+    int i=0;
+
+    if (uname(&buffer) != 0) {
+        return -1;
+    }
+
+    p = buffer.release;
+
+    while (*p) {
+        if (isdigit(*p)) {
+            ver[i] = strtol(p, &p, 10);
+            i++;
+        } else {
+            p++;
+        }
+    }
+
+// FIXME: Should replace these with actual kernel version where multishot poll is integrated, likely 5.13
+    if ((ver[0] > 5) || ((ver[0] == 5) && (ver[1] >= 12))) {
```

Comment:
There are missing bounds checks here, presumably a check should have been done on `i`?